_posts/2018-09-04-http-part4.markdown

```diff
@@ -0,0 +1,236 @@
+
+### File descriptors
+Imagine that database is now taking forever to return query, our requests handling tasks are piling up on a queue. Remember that before we submit a task to thread pool we accept a new connection. Eventually, what will happen is that we will hit a file descriptors limit! As you may remember from [Part 2](http://dyszkiewicz.me/programming/http/server/kotlin/2018/08/11/http-part2.html), every open connection creates new file descriptor. 
+To see that we setup Vagrant VM changing it open files limit to 500
+```
+sudo bash -c "echo '* - nofile 500' >> /etc/security/limits.conf"
+```
+Then we can run [`wrk` - a HTTP benchmarking tool](https://github.com/wg/wrk) from host machine that opens 1000 concurrent connections
+```
+wrk -c1000 -d10s -t1000 http://localhost:8091
+```
+And there we go, we’ve got an exception
+```
+Cvagrant@vagrant-ubuntu-trusty-64:~$ java -jar /project/naphi-0.1-SNAPSHOT.jar
+Exception in thread "main" java.net.SocketException: Too many open files (Accept failed)
+	at java.net.PlainSocketImpl.socketAccept(Native Method)
+	at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:409)
+	at java.net.ServerSocket.implAccept(ServerSocket.java:545)
+	at java.net.ServerSocket.accept(ServerSocket.java:513)
+	at org.naphi.Server.start(Server.kt:79)
+	at org.naphi.ServerKt.main(Server.kt:119)
+```
+
+The default limit on Ubuntu Vagrant was 4096 on Java process.  You can easily bump it up, but keep in mind that every resource is finite. Every open socket cost us RAM, although as I mentioned in previous posts, you can bump this limit to even millions file descriptors with proper tuning and large enough machine.
+
+#### Can we be more graceful?
+When we hit the file descriptor limit, our server either drops the connection or send `RST` packet. A client will most likely retry connection either by custom logic on `RST` or built-in TCP retries on dropped connection. Instead, we can return [`503 Service Unavailable` status code](https://httpstatuses.com/503). What is interesting is that you can attach a `Retry-After` header which will suggest when the client should try again. If the client respects that, then you can avoid unnecessary requests.
+
+### Number of worker threads
+Let’s say that database is taking 5 seconds to return a query. We’ve got 20 worker threads and in every thread, we are blocking for those 5 seconds. Essentially we can now respond with a maximum of 20 requests per 5 seconds = 4 requests/s.
+Assuming we can spawn other connections to the database and we won’t kill it by doing so (not a real world scenario). Theoretically, if we increase our pool to 200 workers we should get 40 requests/s. 
```

Comment:
This sentence seems like it's been a part of previous one and then got splitted, now it sounds a little weird. Please consider joining it or modifying, because for me it sounds like there's a thought you're trying to present here, but it gets interrupted